File: crates/engine/src/insights.rs
//! AI Insights request, response, and settings contracts.

use serde::{Deserialize, Serialize};

/// Stable persisted preference keys for non-secret AI Insights settings.
pub struct AiInsightsSettingsKeys;

impl AiInsightsSettingsKeys {
    pub const AUTH_MODE: &str = "ai_insights_auth_mode";
    pub const CLOUD_PROFILE: &str = "ai_insights_cloud_profile";
    pub const ENDPOINT: &str = "ai_insights_endpoint";
    pub const DEPLOYMENT_NAME: &str = "ai_insights_deployment_name";
    pub const MODEL_NAME: &str = "ai_insights_model_name";
    pub const ALL: [&'static str; 5] = [
        Self::AUTH_MODE,
        Self::CLOUD_PROFILE,
        Self::ENDPOINT,
        Self::DEPLOYMENT_NAME,
        Self::MODEL_NAME,
    ];
}

/// User-selected authentication mode for AI Insights.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AiInsightsAuthMode {
    #[default]
    AzureLogin,
    ApiKey,
}

/// Selected Azure cloud profile for Azure OpenAI requests.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AiInsightsCloudProfile {
    #[default]
    Commercial,
    UsGovernment,
    UsGovernmentSecret,
    UsGovernmentTopSecret,
}

/// Non-secret settings required to configure AI Insights.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsSettings {
    pub auth_mode: AiInsightsAuthMode,
    pub cloud_profile: AiInsightsCloudProfile,
    pub endpoint: String,
    pub deployment_name: String,
    #[serde(default)]
    pub model_name: Option<String>,
}

/// Scope visible to the model contract so namespace-limited requests stay explicit.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", tag = "kind")]
pub enum AiInsightsScope {
    Cluster,
    Namespace { namespace: String },
}

/// Engine-owned request contract for generating AI Insights.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsRequest {
    pub scope: AiInsightsScope,
    pub settings: AiInsightsSettings,
}

/// Stable redaction policy version for AI Insights context shaping.
pub const AI_INSIGHTS_REDACTION_POLICY_VERSION: &str = "v1";

/// Maximum workload summaries included in the shaped AI Insights context.
pub const AI_INSIGHTS_WORKLOAD_CAP: usize = 12;
/// Maximum pod summaries included in the shaped AI Insights context.
pub const AI_INSIGHTS_POD_CAP: usize = 12;
/// Maximum warning-event summaries included in the shaped AI Insights context.
pub const AI_INSIGHTS_EVENT_CAP: usize = 10;
/// Maximum node summaries included in the shaped AI Insights context.
pub const AI_INSIGHTS_NODE_CAP: usize = 8;
/// Maximum Helm release summaries included in the shaped AI Insights context.
pub const AI_INSIGHTS_HELM_RELEASE_CAP: usize = 10;

/// Allowlist-only context contract sent to later AI orchestration.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsContext {
    pub scope: AiInsightsScope,
    pub redaction_policy_version: String,
    pub connection: AiInsightsConnectionSummary,
    pub workloads: AiInsightsCollection<AiInsightsWorkloadSummary>,
    pub pods: AiInsightsCollection<AiInsightsPodSummary>,
    pub events: AiInsightsCollection<AiInsightsEventSummary>,
    pub nodes: AiInsightsCollection<AiInsightsNodeSummary>,
    pub helm_releases: AiInsightsCollection<AiInsightsHelmReleaseSummary>,
    pub aks: Option<AiInsightsAksSummary>,
}

/// Curated connection summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsConnectionSummary {
    pub status: AiInsightsConnectionStatus,
    pub message: Option<String>,
    pub resources_synced: Option<u32>,
    pub resources_total: Option<u32>,
}

/// Connection status values surfaced to AI Insights orchestration.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AiInsightsConnectionStatus {
    Disconnected,
    Connecting,
    Syncing,
    Ready,
    Degraded,
    Error,
    Backoff,
}

/// Deterministically capped collection of curated summaries.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsCollection<T> {
    pub total_count: u32,
    pub items: Vec<T>,
}

/// Curated workload summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsWorkloadSummary {
    pub kind: String,
    pub namespace: String,
    pub name: String,
    pub desired_replicas: Option<u32>,
    pub ready_replicas: Option<u32>,
    pub available_replicas: Option<u32>,
    pub updated_replicas: Option<u32>,
    pub unavailable_replicas: Option<u32>,
    pub issue: Option<String>,
}

/// Curated pod summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsPodSummary {
    pub namespace: String,
    pub name: String,
    pub phase: Option<String>,
    pub ready_containers: u32,
    pub total_containers: u32,
    pub restart_count: u32,
    pub node_name: Option<String>,
    pub issue: Option<String>,
}

/// Curated warning-event summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsEventSummary {
    pub namespace: String,
    pub regarding_kind: String,
    pub regarding_name: String,
    pub reason: String,
    pub message: String,
    pub count: u32,
    pub last_seen: Option<String>,
}

/// Curated node summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsNodeSummary {
    pub name: String,
    pub ready: bool,
    pub unschedulable: bool,
    pub pressures: Vec<String>,
    pub kubelet_version: Option<String>,
}

/// Curated Helm release summary included in the shaped AI Insights context.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsHelmReleaseSummary {
    pub namespace: String,
    pub name: String,
    pub chart: String,
    pub app_version: String,
    pub revision: i32,
    pub status: String,
}

/// Narrow AKS posture summary included when desktop ARM data is available.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsAksSummary {
    pub kubernetes_version: Option<String>,
    pub provisioning_state: Option<String>,
    pub power_state: Option<String>,
    pub dns_prefix: Option<String>,
    pub private_cluster: Option<bool>,
    pub local_accounts_disabled: Option<bool>,
    pub network_plugin: Option<String>,
    pub outbound_type: Option<String>,
}

/// Strict advisory-only response contract rendered by the UI.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsResponse {
    pub summary: String,
    pub risks: Vec<AiInsightsRisk>,
    pub observations: Vec<AiInsightsObservation>,
    pub recommendations: Vec<AiInsightsRecommendation>,
    pub references: Vec<AiInsightsReference>,
}

/// Risk item included in the structured AI Insights response.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsRisk {
    pub title: String,
    pub detail: String,
    pub impact: AiInsightsRiskImpact,
}

/// Supported impact levels for AI Insights risks.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AiInsightsRiskImpact {
    Low,
    Medium,
    High,
}

/// Observation item included in the structured AI Insights response.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsObservation {
    pub area: String,
    pub detail: String,
}

/// Recommendation item included in the structured AI Insights response.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsRecommendation {
    pub action: String,
    pub rationale: String,
    pub confidence: f64,
}

/// Resource reference included in the structured AI Insights response.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsReference {
    pub kind: String,
    pub name: String,
    pub namespace: Option<String>,
}

/// Dev-only diagnostics metadata for Settings.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsDevDiagnostics {
    pub prompt_version: Option<String>,
    pub redaction_policy_version: Option<String>,
    pub cloud_profile: AiInsightsCloudProfile,
    pub auth_mode: AiInsightsAuthMode,
    pub context_size: Option<AiInsightsContextSize>,
    pub schema_validation_failure: Option<AiInsightsSchemaValidationFailure>,
    pub provider_error_classification: Option<AiInsightsProviderErrorClass>,
}

/// Context-size metadata surfaced only in dev mode.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsContextSize {
    pub serialized_bytes: u64,
    pub resource_count: u32,
}

/// Schema-validation metadata surfaced only in dev mode.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AiInsightsSchemaValidationFailure {
    pub path: Option<String>,
    pub message: String,
}

/// Classified provider failure metadata surfaced only in dev mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum AiInsightsProviderErrorClass {
    Configuration,
    Credential,
    Authorization,
    Endpoint,
    Timeout,
    Network,
    Unknown,
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn request_serializes_namespace_scope_explicitly() {
        let request = AiInsightsRequest {
            scope: AiInsightsScope::Namespace {
                namespace: "payments".to_string(),
            },
            settings: AiInsightsSettings {
                auth_mode: AiInsightsAuthMode::ApiKey,
                cloud_profile: AiInsightsCloudProfile::UsGovernment,
                endpoint: "https://example.openai.azure.com".to_string(),
                deployment_name: "gpt-4o-mini".to_string(),
                model_name: Some("gpt-4o-mini".to_string()),
            },
        };

        let value = serde_json::to_value(&request).unwrap();

        assert_eq!(
            value,
            json!({
                "scope": {
                    "kind": "namespace",
                    "namespace": "payments"
                },
                "settings": {
                    "authMode": "apiKey",
                    "cloudProfile": "usGovernment",
                    "endpoint": "https://example.openai.azure.com",
                    "deploymentName": "gpt-4o-mini",
                    "modelName": "gpt-4o-mini"
                }
            })
        );
    }

    #[test]
    fn response_serializes_with_prd_shape() {
        let response = AiInsightsResponse {
            summary: "Cluster health is degraded by a single workload.".to_string(),
            risks: vec![AiInsightsRisk {
                title: "Crash looping deployment".to_string(),
                detail: "payments-api restarted 9 times in the last hour.".to_string(),
                impact: AiInsightsRiskImpact::High,
            }],
            observations: vec![AiInsightsObservation {
                area: "Workloads".to_string(),
                detail: "Replica availability is healthy outside the payments namespace."
                    .to_string(),
            }],
            recommendations: vec![AiInsightsRecommendation {
                action: "Inspect recent events for payments-api".to_string(),
                rationale: "Repeated restarts usually correlate with probe or config failures."
                    .to_string(),
                confidence: 0.92,
            }],
            references: vec![AiInsightsReference {
                kind: "Deployment".to_string(),
                name: "payments-api".to_string(),
                namespace: Some("payments".to_string()),
            }],
        };

        let value = serde_json::to_value(&response).unwrap();

        assert_eq!(
            value,
            json!({
                "summary": "Cluster health is degraded by a single workload.",
                "risks": [
                    {
                        "title": "Crash looping deployment",
                        "detail": "payments-api restarted 9 times in the last hour.",
                        "impact": "high"
                    }
                ],
                "observations": [
                    {
                        "area": "Workloads",
                        "detail": "Replica availability is healthy outside the payments namespace."
                    }
                ],
                "recommendations": [
                    {
                        "action": "Inspect recent events for payments-api",
                        "rationale": "Repeated restarts usually correlate with probe or config failures.",
                        "confidence": 0.92
                    }
                ],
                "references": [
                    {
                        "kind": "Deployment",
                        "name": "payments-api",
                        "namespace": "payments"
                    }
                ]
            })
        );
    }

    #[test]
    fn response_rejects_unknown_fields() {
        let invalid = json!({
            "summary": "ok",
            "risks": [],
            "observations": [],
            "recommendations": [],
            "references": [],
            "chat": "not allowed"
        });

        let result = serde_json::from_value::<AiInsightsResponse>(invalid);

        assert!(result.is_err());
    }

    #[test]
    fn diagnostics_roundtrip_uses_explicit_metadata_fields() {
        let diagnostics = AiInsightsDevDiagnostics {
            prompt_version: Some("2026-03-24".to_string()),
            redaction_policy_version: Some("v1".to_string()),
            cloud_profile: AiInsightsCloudProfile::Commercial,
            auth_mode: AiInsightsAuthMode::AzureLogin,
            context_size: Some(AiInsightsContextSize {
                serialized_bytes: 2048,
                resource_count: 17,
            }),
            schema_validation_failure: Some(AiInsightsSchemaValidationFailure {
                path: Some("recommendations[0].confidence".to_string()),
                message: "confidence must be between 0 and 1".to_string(),
            }),
            provider_error_classification: Some(AiInsightsProviderErrorClass::Authorization),
        };

        let json = serde_json::to_string(&diagnostics).unwrap();
        let roundtrip: AiInsightsDevDiagnostics = serde_json::from_str(&json).unwrap();

        assert_eq!(roundtrip, diagnostics);
    }

    #[test]
    fn context_serializes_with_explicit_scope_and_curated_categories() {
        let context = AiInsightsContext {
            scope: AiInsightsScope::Namespace {
                namespace: "payments".to_string(),
            },
            redaction_policy_version: AI_INSIGHTS_REDACTION_POLICY_VERSION.to_string(),
            connection: AiInsightsConnectionSummary {
                status: AiInsightsConnectionStatus::Ready,
                message: None,
                resources_synced: Some(29),
                resources_total: Some(29),
            },
            workloads: AiInsightsCollection {
                total_count: 1,
                items: vec![AiInsightsWorkloadSummary {
                    kind: "Deployment".to_string(),
                    namespace: "payments".to_string(),
                    name: "payments-api".to_string(),
                    desired_replicas: Some(3),
                    ready_replicas: Some(2),
                    available_replicas: Some(2),
                    updated_replicas: Some(3),
                    unavailable_replicas: Some(1),
                    issue: Some("1 replica is unavailable".to_string()),
                }],
            },
            pods: AiInsightsCollection {
                total_count: 0,
                items: Vec::new(),
            },
            events: AiInsightsCollection {
                total_count: 0,
                items: Vec::new(),
            },
            nodes: AiInsightsCollection {
                total_count: 0,
                items: Vec::new(),
            },
            helm_releases: AiInsightsCollection {
                total_count: 0,
                items: Vec::new(),
            },
            aks: None,
        };

        let value = serde_json::to_value(&context).unwrap();

        assert_eq!(value["scope"]["kind"], "namespace");
        assert_eq!(value["scope"]["namespace"], "payments");
        assert_eq!(value["redactionPolicyVersion"], "v1");
        assert!(value.get("workloads").is_some());
        assert!(value.get("pods").is_some());
        assert!(value.get("events").is_some());
        assert!(value.get("nodes").is_some());
        assert!(value.get("helmReleases").is_some());
        assert!(value.get("aks").is_some());
    }

    #[test]
    fn settings_keys_are_stable_and_complete() {
        assert_eq!(
            AiInsightsSettingsKeys::ALL,
            [
                "ai_insights_auth_mode",
                "ai_insights_cloud_profile",
                "ai_insights_endpoint",
                "ai_insights_deployment_name",
                "ai_insights_model_name",
            ]
        );
    }
}

File: apps/web/src/lib/insights.ts
export const AI_INSIGHTS_AUTH_MODES = ['azureLogin', 'apiKey'] as const;

export type AiInsightsAuthMode = (typeof AI_INSIGHTS_AUTH_MODES)[number];

export const AI_INSIGHTS_CLOUD_PROFILES = [
  'commercial',
  'usGovernment',
  'usGovernmentSecret',
  'usGovernmentTopSecret',
] as const;

export type AiInsightsCloudProfile = (typeof AI_INSIGHTS_CLOUD_PROFILES)[number];
export type AzureCloudProfile = AiInsightsCloudProfile;

export const AZURE_CLOUD_PROFILES = AI_INSIGHTS_CLOUD_PROFILES;

export const AI_INSIGHTS_PROVIDER_ERROR_CLASSES = [
  'configuration',
  'credential',
  'authorization',
  'endpoint',
  'timeout',
  'network',
  'unknown',
] as const;

export type AiInsightsProviderErrorClass =
  (typeof AI_INSIGHTS_PROVIDER_ERROR_CLASSES)[number];

export type AiInsightsRiskImpact = 'low' | 'medium' | 'high';

export interface AiInsightsRisk {
  title: string;
  detail: string;
  impact: AiInsightsRiskImpact;
}

export interface AiInsightsObservation {
  area: string;
  detail: string;
}

export interface AiInsightsRecommendation {
  action: string;
  rationale: string;
  confidence: number;
}

export interface AiInsightsReference {
  kind: string;
  name: string;
  namespace: string | null;
}

export interface AiInsightsResponse {
  summary: string;
  risks: AiInsightsRisk[];
  observations: AiInsightsObservation[];
  recommendations: AiInsightsRecommendation[];
  references: AiInsightsReference[];
}

export interface AiInsightsConnectionTestResult {
  normalizedEndpoint: string;
  chatCompletionsUrl: string;
  model: string;
}

export type AiInsightsScope =
  | { kind: 'cluster' }
  | { kind: 'namespace'; namespace: string };

export interface AiInsightsHistoryEntry {
  createdAt: string;
  scope: AiInsightsScope;
  response: AiInsightsResponse;
}

export interface AiInsightsSettings {
  endpoint: string;
  deploymentName: string;
  authMode: AiInsightsAuthMode;
  cloudProfile: AiInsightsCloudProfile;
  modelName: string | null;
}

export interface AiInsightsDevDiagnostics {
  promptVersion?: string | null;
  redactionPolicyVersion?: string | null;
  cloudProfile: AiInsightsCloudProfile;
  authMode: AiInsightsAuthMode;
  contextSize?: AiInsightsContextSize | null;
  schemaValidationFailure?: AiInsightsSchemaValidationFailure | null;
  providerErrorClassification?: AiInsightsProviderErrorClass | null;
}

export const AI_INSIGHTS_SETTINGS_KEYS = {
  endpoint: 'ai_insights_endpoint',
  deploymentName: 'ai_insights_deployment_name',
  authMode: 'ai_insights_auth_mode',
  cloudProfile: 'ai_insights_cloud_profile',
  modelName: 'ai_insights_model_name',
} as const;

export type AiInsightsSettingsKey =
  (typeof AI_INSIGHTS_SETTINGS_KEYS)[keyof typeof AI_INSIGHTS_SETTINGS_KEYS];

export const DEFAULT_AI_INSIGHTS_AUTH_MODE: AiInsightsAuthMode = 'azureLogin';
export const DEFAULT_AZURE_CLOUD_PROFILE: AiInsightsCloudProfile = 'commercial';

export interface AiInsightsContextSize {
  serializedBytes: number;
  resourceCount: number;
}

export interface AiInsightsSchemaValidationFailure {
  path?: string | null;
  message: string;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasOnlyKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  return Object.keys(value).every((key) => keys.includes(key));
}

function hasExactKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  return Object.keys(value).length === keys.length && hasOnlyKeys(value, keys);
}

function isOptionalString(value: unknown): value is string | null | undefined {
  return value === undefined || value === null || typeof value === 'string';
}

function isNonNegativeInteger(value: unknown): value is number {
  return typeof value === 'number' && Number.isInteger(value) && value >= 0;
}

export function isAiInsightsAuthMode(value: unknown): value is AiInsightsAuthMode {
  return typeof value === 'string'
    && (AI_INSIGHTS_AUTH_MODES as readonly string[]).includes(value);
}

export function isAiInsightsCloudProfile(value: unknown): value is AiInsightsCloudProfile {
  return typeof value === 'string'
    && (AI_INSIGHTS_CLOUD_PROFILES as readonly string[]).includes(value);
}

export function isAzureCloudProfile(value: unknown): value is AzureCloudProfile {
  return isAiInsightsCloudProfile(value);
}

export function isAiInsightsProviderErrorClass(value: unknown): value is AiInsightsProviderErrorClass {
  return typeof value === 'string'
    && (AI_INSIGHTS_PROVIDER_ERROR_CLASSES as readonly string[]).includes(value);
}

export function parseAiInsightsAuthMode(value: string | null | undefined): AiInsightsAuthMode {
  return isAiInsightsAuthMode(value) ? value : DEFAULT_AI_INSIGHTS_AUTH_MODE;
}

export function parseAiInsightsCloudProfile(
  value: string | null | undefined,
): AiInsightsCloudProfile {
  return isAiInsightsCloudProfile(value) ? value : DEFAULT_AZURE_CLOUD_PROFILE;
}

export function parseAzureCloudProfile(value: string | null | undefined): AzureCloudProfile {
  return parseAiInsightsCloudProfile(value);
}

export function formatAiInsightsScope(scope: AiInsightsScope): string {
  return scope.kind === 'cluster' ? 'cluster' : `namespace/${scope.namespace}`;
}

export function createDefaultAiInsightsSettings(
  cloudProfile: AiInsightsCloudProfile = DEFAULT_AZURE_CLOUD_PROFILE,
): AiInsightsSettings {
  return {
    endpoint: '',
    deploymentName: '',
    authMode: DEFAULT_AI_INSIGHTS_AUTH_MODE,
    cloudProfile,
    modelName: null,
  };
}

export function isAiInsightsResponse(value: unknown): value is AiInsightsResponse {
  if (!isRecord(value)
    || !hasExactKeys(value, ['summary', 'risks', 'observations', 'recommendations', 'references'])
    || typeof value.summary !== 'string') {
    return false;
  }

  const { risks, observations, recommendations, references } = value;

  return Array.isArray(risks)
    && risks.every((risk) => isRecord(risk)
      && hasExactKeys(risk, ['title', 'detail', 'impact'])
      && typeof risk.title === 'string'
      && typeof risk.detail === 'string'
      && (risk.impact === 'low' || risk.impact === 'medium' || risk.impact === 'high'))
    && Array.isArray(observations)
    && observations.every((observation) => isRecord(observation)
      && hasExactKeys(observation, ['area', 'detail'])
      && typeof observation.area === 'string'
      && typeof observation.detail === 'string')
    && Array.isArray(recommendations)
    && recommendations.every((recommendation) => isRecord(recommendation)
      && hasExactKeys(recommendation, ['action', 'rationale', 'confidence'])
      && typeof recommendation.action === 'string'
      && typeof recommendation.rationale === 'string'
      && typeof recommendation.confidence === 'number'
      && Number.isFinite(recommendation.confidence))
    && Array.isArray(references)
    && references.every((reference) => isRecord(reference)
      && hasExactKeys(reference, ['kind', 'name', 'namespace'])
      && typeof reference.kind === 'string'
      && typeof reference.name === 'string'
      && (typeof reference.namespace === 'string' || reference.namespace === null));
}

export function isAiInsightsScope(value: unknown): value is AiInsightsScope {
  if (!isRecord(value) || typeof value.kind !== 'string') {
    return false;
  }

  if (value.kind === 'cluster') {
    return hasExactKeys(value, ['kind']);
  }

  if (value.kind === 'namespace') {
    return hasExactKeys(value, ['kind', 'namespace'])
      && typeof value.namespace === 'string';
  }

  return false;
}

export function isAiInsightsConnectionTestResult(
  value: unknown,
): value is AiInsightsConnectionTestResult {
  return isRecord(value)
    && hasExactKeys(value, ['normalizedEndpoint', 'chatCompletionsUrl', 'model'])
    && typeof value.normalizedEndpoint === 'string'
    && typeof value.chatCompletionsUrl === 'string'
    && typeof value.model === 'string';
}

export function isAiInsightsHistoryEntry(value: unknown): value is AiInsightsHistoryEntry {
  return isRecord(value)
    && hasExactKeys(value, ['createdAt', 'scope', 'response'])
    && typeof value.createdAt === 'string'
    && isAiInsightsScope(value.scope)
    && isAiInsightsResponse(value.response);
}

function isAiInsightsContextSize(value: unknown): value is AiInsightsContextSize {
  return isRecord(value)
    && hasExactKeys(value, ['serializedBytes', 'resourceCount'])
    && isNonNegativeInteger(value.serializedBytes)
    && isNonNegativeInteger(value.resourceCount);
}

function isAiInsightsSchemaValidationFailure(
  value: unknown,
): value is AiInsightsSchemaValidationFailure {
  return isRecord(value)
    && hasOnlyKeys(value, ['path', 'message'])
    && typeof value.message === 'string'
    && isOptionalString(value.path);
}

export function isAiInsightsDevDiagnostics(value: unknown): value is AiInsightsDevDiagnostics {
  return isRecord(value)
    && hasOnlyKeys(value, [
      'promptVersion',
      'redactionPolicyVersion',
      'cloudProfile',
      'authMode',
      'contextSize',
      'schemaValidationFailure',
      'providerErrorClassification',
    ])
    && isOptionalString(value.promptVersion)
    && isOptionalString(value.redactionPolicyVersion)
    && isAiInsightsCloudProfile(value.cloudProfile)
    && isAiInsightsAuthMode(value.authMode)
    && (value.contextSize === undefined
      || value.contextSize === null
      || isAiInsightsContextSize(value.contextSize))
    && (value.schemaValidationFailure === undefined
      || value.schemaValidationFailure === null
      || isAiInsightsSchemaValidationFailure(value.schemaValidationFailure))
    && (value.providerErrorClassification === undefined
      || value.providerErrorClassification === null
      || isAiInsightsProviderErrorClass(value.providerErrorClassification));
}
